internal/gc.h

```diff
@@ -141,25 +142,14 @@ struct rb_objspace; /* in vm_core.h */
 # undef RB_NEWOBJ_OF
 #endif
 
-#define RVALUE_SIZE (sizeof(struct RBasic) + sizeof(VALUE[RBIMPL_RVALUE_EMBED_LEN_MAX]))
-
-#define RB_RVARGC_NEWOBJ_OF(var, T, c, f, s) \
-  T *(var) = (T *)(((f) & FL_WB_PROTECTED) ? \
-                   rb_wb_protected_newobj_of((c), (f) & ~FL_WB_PROTECTED, s) : \
-                   rb_wb_unprotected_newobj_of((c), (f), s))
-
-#define RB_RVARGC_EC_NEWOBJ_OF(ec, var, T, c, f, s) \
+#define RVARGC_NEWOBJ_OF(var, T, c, f, s, ec) \
   T *(var) = (T *)(((f) & FL_WB_PROTECTED) ? \
-                   rb_ec_wb_protected_newobj_of((ec), (c), (f) & ~FL_WB_PROTECTED, s) : \
+                   rb_wb_protected_newobj_of((ec) ? (ec) : GET_EC(), (c), (f) & ~FL_WB_PROTECTED, s) : \
                    rb_wb_unprotected_newobj_of((c), (f), s))
 
-/* optimized version of NEWOBJ() */
-#define RB_NEWOBJ_OF(var, T, c, f) RB_RVARGC_NEWOBJ_OF(var, T, c, f, RVALUE_SIZE)
-
-#define RB_EC_NEWOBJ_OF(ec, var, T, c, f) RB_RVARGC_EC_NEWOBJ_OF(ec, var, T, c, f, RVALUE_SIZE)
-
+#define RB_NEWOBJ_OF(var, T, c, f) RVARGC_NEWOBJ_OF(var, T, c, f, RVALUE_SIZE, NULL)
 #define NEWOBJ_OF(var, T, c, f) RB_NEWOBJ_OF((var), T, (c), (f))
```

Comment:
Perhaps out of scope of this PR, but it might be nice to get rid of these two macros since we can essentially use `RVARGC_NEWOBJ_OF` (and pass in the size of the type we're trying to allocate). This might be beneficial for some allocators since not all objects are of `RVALUE_SIZE`.

Reply:
Agreed. I didn't attempt this for this PR because these macros are also defined in the public `newobj.h` header. I'll likely attempt to remove/deprecate these in a future PR.

Reply:
Isn't that a reason to remove this internally first? This means that there's a name conflict with `newobj.h` which will result in behavior that differs depending on the load order.